### PR TITLE
Verify import aliases

### PR DESCRIPTION
**Pull request by dims**
- Added scripts for update and verify
- golang AST code for scanning and fixing imports
- default regex allows it to run on just test/e2e.* file paths
- exclude verify-import-aliases.sh from running in CI jobs

Co-Authored-By: Aaron Crickenberger <spiffxp@google.com>

### CODE REVIEW

**Comment by dims** [comment]
Looks like a good start https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/78836/pull-kubernetes-typecheck/1137882903812771841

**Comment by draveness** [comment]
This is Great! 😁

**Comment by johnSchnake** [comment]
I'd be fine with this but something that effects everyone wouldn't be my say. I do like that it is not tied to a particular tool (e.g. goimports) but enforces a standard via automation.

Would this just check the files the user changed or is this going to cause breakages until every file adheres to the alias list?

**Comment by draveness** [comment]
> Would this check the files the user changed or is this going to cause breakages until every file adheres to the alias list?

We could make this happen by checking the diff with the master branch, and it would be much easier to obey.

**Comment by dims** [comment]
@johnSchnake `until every file adheres to the alias list` with the code as-is

@johnSchnake @draveness feel free to take this over :)

dims [11:32 AM]
@jschnake wanna pick that up, tidy it to unblock your and @spiffxp’s ideas/prs?
- you could also chat with @rmmh about whether it would be better to have a separate file (copy of test/typecheck/main.go, but possibly stripped down) or leave it where it is
- we may have to reduce the duplicate lines of errors, message in the verify-typecheck.sh etc as well.

File: cmd/preferredimports/preferredimports.go
Comment (draveness) [inline cmd/preferredimports/preferredimports.go]: it will be a lot of errors if scan all the files in k/k when we add a new rule. We could have something like this:

```go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)

	files := []string{}
	for {
		file, err := reader.ReadString('\n')
		if err != nil {
			if err == io.EOF {
				break
			}

			panic(err)
		}

		files = append(files, file)
	}

	fmt.Println(files)
}

```

File: hack/verify-import-aliases.sh
Comment (draveness) [inline hack/verify-import-aliases.sh]: git diff --name-only master | go run cmd/preferredimports/preferredimports.go "$@" || ret=$?

File: cmd/preferredimports/preferredimports.go
Comment (draveness) [inline cmd/preferredimports/preferredimports.go]: an ignore liked file would be better

```go
.*
staging/
vendor/
_output/
pkg/kubectl/cmd/testdata/edit
```

**Comment by liggitt** [comment]
if we're going to add import alias restrictions like this, I think we need a paired fixup script (like verify-gofmt / update-gofmt). The added toil to round-trip running verify and manually fixing up files to satisfy a bot outweighs the benefits of import aliases in my opinion.

File: cmd/preferredimports/preferredimports.go
Comment (dims) [reply cmd/preferredimports/preferredimports.go]: I don't want to optimize this right now as the time taken is pretty small.

File: hack/verify-import-aliases.sh
Comment (dims) [reply hack/verify-import-aliases.sh]: If the idea is to run the script only for the current changes, there's another way to do it already

```
git diff --name-only "$(git merge-base HEAD master)"...HEAD | sed 's|\(.*\)/.*|\1|' | sort | uniq | xargs go run cmd/preferredimports/preferredimports.go
```

File: cmd/preferredimports/preferredimports.go
Comment (dims) [reply cmd/preferredimports/preferredimports.go]: Right, i *really* don't want to add more files, so having it in code forces that.

**Comment by dims** [comment]
Ack @liggitt

File: hack/verify-import-aliases.sh
Comment (liggitt) [reply hack/verify-import-aliases.sh]: That will force people to update aliases as they touch existing files. Similar things happen with lint occasionally (the lint tool will start complaining about a file that is passing lint checks on master, and require fixup in an unrelated PR). Failing CI on changes unrelated to a PR is probably friction we don't need.

Comment (dims) [reply hack/verify-import-aliases.sh]: Agree, that's why i was running on the whole code base and not limiting to the current change

Comment (draveness) [reply hack/verify-import-aliases.sh]: Does it mean that if we want to add a new rule, we have to fix all of the import aliases in the whole codebase? If yes, then we have to work on a `update-import-aliases` script, which could always help us to fix up the code.

Comment (dims) [reply hack/verify-import-aliases.sh]: Yes

**Comment by dims** [comment]
xref https://github.com/kubernetes/kubernetes/pull/78780
xref https://github.com/kubernetes/kubernetes/pull/76863
xref https://github.com/kubernetes/kubernetes/pull/77407

File: cmd/preferredimports/preferredimports.go
Comment (johnSchnake) [inline cmd/preferredimports/preferredimports.go]: nit; would prefer this go to `stderr` so prow recognizes it as part of the failure messages. Otherwise it tells you an error occurred and to check the preferred output, but you have to download and grep the whole log to get which errors you encountered.

File: cmd/preferredimports/preferredimports.go
Comment (johnSchnake) [inline cmd/preferredimports/preferredimports.go]: nit, prefer `Fprintf` to os.Stderr (not going to mention this again, but the point is the same for each of the error messages)

File: hack/update-import-aliases.sh
Comment (johnSchnake) [inline hack/update-import-aliases.sh]: minor nit; i know lots of the script was probably copy/pasted, but should you update the year on a number of these files to be 2019? Whats the protocol on when to update it if you are, for instance, copy/pasting/editing an existing script?

**Review by johnSchnake — COMMENTED** [review]
For what it is, lgtm, but I'm still a little fuzzy about the upcoming use. It still seems like we aren't finding a full consensus so confirm if my current understanding is correct:
 - this PR is adding the tool/script to do find and fix the problems
 - follow up PRs will run this and update the files across the code base (or a subset? what subset?)
 - no changes to CI yet regarding this and causing failures? Is that the plan eventually?

File: cmd/preferredimports/preferredimports.go
Comment (dims) [reply cmd/preferredimports/preferredimports.go]: Fixed

File: cmd/preferredimports/preferredimports.go
Comment (dims) [reply cmd/preferredimports/preferredimports.go]: Fixed

File: hack/update-import-aliases.sh
Comment (dims) [reply hack/update-import-aliases.sh]: Fixed

**Comment by dims** [comment]
@draveness @spiffxp @johnSchnake Can you please try doing what you needed in #78780 / #76863 / #77407 by adding the import alias in `hack/.import-aliases` and running the update/verify scripts?

@cblecker this is ready for a review. PTAL

**Comment by dims** [comment]
Based on the discussion, what we have now here:
- we have a AST based golang code, it's *FAST*
- we have both update and verify scripts
- defaults to running with regex `(test/e2e|test/e2e_node)`
- we can ask folks to add what they need in `hack/.import-aliases` for the aliases, i have a sample commit for one of the aliases from a previous PR
- scripts can run on arbitrary directories within the regex if the user chooses to.
- does not limit itself to GOOS/GOARCH on which the script is executed.

@liggitt Any other concerns?

File: hack/.import-aliases
Comment (liggitt) [inline hack/.import-aliases]: can you describe how this file is expected to be maintained? a manual mapping like this will drift and be subject to typos, right?

**Comment by dims** [comment]
as discussed in the k8s-code-organization meeting, let's not run the verify job in the CI. folks working on conformance can run it periodically to keep the imports sane.

Thanks,
Dims

**Comment by dims** [comment]
@spiffxp @johnSchnake please revive your PRs that fixup imports once this lands, please edit the hack/.import-aliases file and run `hack/update-import-aliases.sh` (and use the verify script to sniff test as well)

File: cmd/preferredimports/preferredimports.go
Comment (spiffxp) [inline cmd/preferredimports/preferredimports.go]: I wouldn't touch test/e2e_kubeadm just yet
```suggestion
	regex         = flag.String("include-path", "(test/e2e/|test/e2e_node)", "only files with paths matching this regex is touched")
```

**Review by spiffxp — CHANGES_REQUESTED** [review]
working on redoing my changes via this script

File: hack/update-import-aliases.sh
Comment (oomichi) [inline hack/update-import-aliases.sh]: This `confirm` operation is for updating go files based on hack/.import-aliases.
So if error happens here, that can be a different reason from these go files import unpreferred aliases.
Better to output different error message.

**Review by oomichi — COMMENTED** [review]
This is a very nice idea for keeping the code consistent, thanks @dims.

File: cmd/preferredimports/preferredimports.go
Comment (spiffxp) [inline cmd/preferredimports/preferredimports.go]: avoid panics on unaliased imports
```suggestion
						if imp.Name != nil {
							imp.Name.Name = alias
						} else {
							imp.Name = ast.NewIdent(alias)
						}
```

File: cmd/preferredimports/preferredimports.go
Comment (spiffxp) [inline cmd/preferredimports/preferredimports.go]: usages of imports without aliases won't get renamed unless we know what the unaliased name is
```suggestion
				pathSegments := strings.Split(importPath, "/")
				importName := pathSegments[len(pathSegments) - 1]
```

File: cmd/preferredimports/preferredimports.go
Comment (spiffxp) [inline cmd/preferredimports/preferredimports.go]: Can't do a multi-line suggestion: we want to avoid renaming package declarations that happen to have the same name as an import alias (eg: `test/e2e/framework/apps/*.go` are in an `apps` package, and also `import apps "k8s.io/api/apps/v1"`)
```go
		switch id := node.(type) {
		case *ast.Package:
			return false
		case *ast.Ident:
			if id.Name == old {
				id.Name = new
			}
		}
```

**Review by spiffxp — COMMENTED** [review]
I'm getting close to calling this too fiddly to be worth our time, but going to try for a bit longer

File: cmd/preferredimports/preferredimports.go
Comment (spiffxp) [reply cmd/preferredimports/preferredimports.go]: n/m this didn't quite seem to fix it

Comment (dims) [reply cmd/preferredimports/preferredimports.go]: @spiffxp can you give me a specific example entry in `hack/.import-aliases` that does not work right?

Comment (spiffxp) [reply cmd/preferredimports/preferredimports.go]: `"k8s.io/api/apps/v1": "appsv1"`

Comment (spiffxp) [reply cmd/preferredimports/preferredimports.go]:
```go
	// use this to avoid renaming the package declaration, eg:
	//   given: package foo; import foo "bar"; foo.Baz, rename foo->qux
	//   yield: package foo; import qux "bar"; qux.Baz
	var pkg *ast.Ident

	// Rename top-level old to new, both unresolved names
	// (probably defined in another file) and names that resolve
	// to a declaration we renamed.
	ast.Inspect(f, func(node ast.Node) bool {
		if node == nil {
			return false
		}
		switch id := node.(type) {
		case *ast.File:
			pkg = id.Name
		case *ast.Ident:
			if pkg != nil && id == pkg {
				return false
			}
			if id.Name == old {
				id.Name = new
			}
		}
		return true
	})
```

File: cmd/preferredimports/preferredimports.go
Comment (spiffxp) [inline cmd/preferredimports/preferredimports.go]: I would move this into the above `if !*confirm` block; in confirm mode we're correcting the error, which I would count as success

File: cmd/preferredimports/preferredimports.go
Comment (dims) [reply cmd/preferredimports/preferredimports.go]: Fixed.! thanks

File: cmd/preferredimports/preferredimports.go
Comment (dims) [reply cmd/preferredimports/preferredimports.go]: Applied the fix

File: cmd/preferredimports/preferredimports.go
Comment (dims) [reply cmd/preferredimports/preferredimports.go]: Done

File: cmd/preferredimports/preferredimports.go
Comment (dims) [reply cmd/preferredimports/preferredimports.go]: Ack applying your fix

**Review by spiffxp — APPROVED** [review]
/lgtm
/approve
for hack/ and test/, I don't have rights for cmd/

**Comment by thockin** [comment]
Why?

What is the problem that this solves?

**Comment by dims** [comment]
@thockin, please see xref #78780 #76863 #77407 and discussions on slack (code org channel). 

We want to make life easier for folks who are working on (or reviewing) conformance and e2e tests by standardizing the imports, it's easy to figure out what alias correspond to which package.

**Comment by spiffxp** [comment]
One other convention this could be used to enforce or help with is the `e2efoo "k8s.io/kubernetes/test/e2e/framework/foo"` convention I see evolving via PRs implementing https://github.com/kubernetes/kubernetes/issues/77095

**Comment by oomichi** [comment]
Thanks for updating,

/lgtm

**Comment by spiffxp** [comment]
Checking back in on this a month later.  126 PRs have merged that touched `test/`, which I used this tool against via https://github.com/kubernetes/kubernetes/pull/78780.  One file needs to be touched up (which I won't bother doing right now)
```
$ ./hack/verify-import-aliases.sh
checking-imports:
ERROR wrong alias for import "k8s.io/api/apps/v1" should be appsv1 in file test/e2e/network/dual_stack.go
```

I regenerated `hack/.import-aliases` using https://gist.github.com/spiffxp/e278d28755a3f64899aaed8f73c78ecf to see if there was any drift.
```diff
--- a/hack/.import-aliases
+++ b/hack/.import-aliases
@@ -35,6 +35,7 @@
   "k8s.io/api/storage/v1": "storagev1",
   "k8s.io/api/storage/v1alpha1": "storagev1alpha1",
   "k8s.io/api/storage/v1beta1": "storagev1beta1",
+  "k8s.io/api/testdata/v1.15.0": "testdatav1.15.0",
   "k8s.io/kubernetes/pkg/controller/apis/config/v1alpha1": "controllerconfigv1alpha1",
   "k8s.io/kubernetes/pkg/kubelet/apis/config/v1beta1": "kubeletconfigv1beta1",
   "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha": "kubeletdevicepluginv1alpha",
@@ -46,6 +47,5 @@
   "k8s.io/kubernetes/pkg/kubelet/apis/resourcemetrics/v1alpha1": "kubeletresourcemetricsv1alpha1",
   "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1": "kubeletstatsv1alpha1",
   "k8s.io/kubernetes/pkg/proxy/apis/config/v1alpha1": "proxyconfigv1alpha1",
-  "k8s.io/kubernetes/pkg/scheduler/apis/config/v1alpha1": "schedulerconfigv1alpha1",
-  "k8s.io/kubernetes/test/e2e/framework/service": "e2eservice"
-}
+  "k8s.io/kubernetes/pkg/scheduler/apis/config/v1alpha1": "schedulerconfigv1alpha1"
+}
```

One manually added e2e entry, and more testdata.  I've modified the script to add those for now

**Pull request merged**
